Feature/edit presentations

natyork opened the pull request:
https://www.pivotaltracker.com/story/show/140277883

This PR updates presentation controller and associated views to allow an admin to edit/update, or delete a presentation.

This PR is dependent on https://github.com/StemboltHQ/staff_tracker/tree/feature/sign-in-mandatory as an admin user needs to be logged in to have update/destroy authorization.

natyork commented:
Stand-by for review while I fix rubocop offenses.

natyork commented:
Rubocop offenses fixeed.

jacquesporveau commented:
Talking with Gray we've decided to move away from a mandatory sign in. I am pretty sure that this PR will solve your problems but perhaps take a look

https://github.com/StemboltHQ/staff_tracker/pull/40

jacquesporveau commented on app/views/presentations/edit.html.erb:
```diff
@@ -0,0 +1,5 @@
+
+<%= render 'form' %>
+
+<%= link_to 'Back', presentations_path %>
```
Kind of hesitant to assume that we know where the user is coming from. There is nothing to stop the user from just typing `presentations/1/edit` in their address bar. This back button would not take them to where they came from in that instance.

natyork replied:
Good point! I used the the previously created people index view as a placeholder to give us easier  access to the edit form and delete link. We may want to add a new story to pivotal to format and style desired views.  I have removed the Back link.

jacquesporveau commented on spec/controllers/presentations_controller_spec.rb:
```diff
@@ -90,6 +90,17 @@
         end
       end
     end
+
```
Good job on the specs.

jacquesporveau commented on app/views/presentations/index.html.erb:
```diff
@@ -0,0 +1,29 @@
+  <thead>
+    <tr>
+      <th>Presentation</th>
+      <th>
```
Missing closing tag.

natyork commented:
Added spec for non-admin trying to GET #edit

jacquesporveau commented on app/views/presentations/index.html.erb:
```diff
@@ -0,0 +1,29 @@
+          <%= link_to 'Edit', edit_presentation_path(presentation) %>
+          <%= link_to 'Delete', presentation_path(presentation),
+                      method: :delete,
+                      data: { confirm: 'Are you sure?' } %>
```
Three links in one table cell seems weird. Interested to see what this looks like.

natyork replied:
It just puts all three links in one line. It doesn't look really great, and I'm not sure that we should be using tables at all, but as mentioned above, I used the people index view as a template for the time being until we have decided on format, flow, and styling for views.

brendandeere replied:
yeah, I'm ok with this. Alternatively we only have links to show and destroy here, and edit is linked to from the show page.

natyork commented:
Updated presentations index view per @jacquesporveau comments and rebased commits onto updated master that merged #40

brendandeere commented on app/controllers/presentations_controller.rb:
```diff
@@ -37,6 +37,14 @@ def update
+  def destroy
+    @presentation = Presentation.find(params[:id])
+    authorize @presentation
+    @presentation.destroy
```
if you are not checking the return value of an AR CRUD method, you should probably use its bang method instead.

`destroy!` will raise an error if the record fails to save.

jacquesporveau replied:
Did you mean it raise an error if it fails to destroy?

brendandeere replied:
Yes. All of the AR Crud methods have a "bang method" equivalent 

so `create` and `create!` 

The normal method returns true or false, the bang method raises an error. Or convention generally is if you are not checking the output of the method, you should raise an error if it fails 

so: 
```ruby
if thing.create
  stuff
else 
  other stuff
end 

# OR 

thing.create! 
```

brendandeere commented on spec/controllers/presentations_controller_spec.rb:
```diff
@@ -90,6 +90,17 @@
     end
+
+    describe 'DELETE #destroy' do
+      let!(:presentation) { FactoryGirl.create(:presentation) }
```
doesnt need to be a `let!` you call this method in the subject so the presentation will get initialised then

natyork replied:
@brendandeere The way this test is implemented it doesn't work without the bang because when presentation gets called in the subject block it immediately destroys the presentation, and so the test to check if the presentation count has changed will fail. Is there a better approach overall for this test?

brendandeere replied:
Right! Given that, you can ignore my previous comment :)

brendandeere commented on app/controllers/presentations_controller.rb:
```diff
@@ -1,4 +1,8 @@
 class PresentationsController < ApplicationController
+  def index
+    @presentation = Presentation.all
```
Not semantic. Presentation implies one record

brendandeere commented on app/views/presentations/index.html.erb:
```diff
@@ -0,0 +1,27 @@
+    </tr>
+  </thead>
+  <tbody>
+    <% @presentation.each do |presentation| %>
```
`<%= render @presentations %>`

natyork commented:
Updated PR per @brendandeere comments and rebased onto updated master that merged #44. I kept the show/edit/delete links on the presentations#index page but also added edit/delete links to the presentations#show view.

dangerdogz commented on app/controllers/presentations_controller.rb:
```diff
@@ -20,6 +29,26 @@ def create
+  def destroy
+    @presentation = Presentation.find(params[:id])
+    authorize @presentation
+    @presentation.destroy
```
As Brendan mentioned, this should be `destroy!` since the return value is not used

natyork replied:
Good catch @dangerdogz, I thought I had updated that. Thanks!

natyork commented:
Updated  presentations#index view to be consistent with events#index.

natyork commented:
Merging #43 into master has caused some functionality conflicts WRT an admin user being able to edit the presenter. Will update the controller & form partial accordingly.

brendandeere commented on app/controllers/presentations_controller.rb:
```diff
@@ -9,9 +9,15 @@ def new
+      @presentation = Presentation.new(admin_presentation_params)
+    else
+      @presentation = Presentation.new(presentation_params)
+                                  .tap do |presentation|
```
I dont think #tap is needed here. I believe you can pass a block directly to the #new method, which will yield the new instance.

Alternatively you might be able to the presentation params to add in current_person. 
`Presentation.new(person: current_person, **presentation_params)` 

Or you could also call merge on the presentation params hash 
`Presentation.new(presentation_params.merge(person: current_person))`

brendandeere commented on spec/controllers/presentations_controller_spec.rb:
```diff
@@ -164,6 +198,18 @@
       end
     end
+
+    describe 'PATCH #update' do
```
Generally we have 1 describe block for each methods and then nest the contexts below that. Here we have the reverse, The contexts outside the describe blocks

natyork replied:
Cool, that makes sense to me! I spent a while trying to rearrange a lot of these methods to fit into the pattern we already had for the project, but I think it would make it much more organized to do it the way you described! I will refactor and update.

brendandeere commented on app/views/presentations/new.html.erb:
```diff
@@ -1,3 +1,3 @@
 <h1>New Presentation</h1>
 
-<%= render 'form' %>
+<%= render partial: 'form', locals: { presentation: @presentation } %>
```
This is older rails syntax. You can pass locals in without the keyword args.

`<%= render 'form', presentation: @presentation %>`

jacquesporveau replied:
Nice to know!

jacquesporveau commented on app/models/presentation.rb:
```diff
@@ -26,6 +26,7 @@ def person_or_presenter
   end
 
   def associated_event_has_available_time
+    return unless duration
```
Two things.

1.) We could avoid the `unless` keyword by going `return if duration.nil?` which is a bit more readable

2.) Duration should probably be a required field in the schema. I imagine if someone is creating a presentation we should be ensuring that it has a duration. That is unrelated to this PR and would be feature creep so this is fine but I'll talk it over with @brendandeere and maybe make a resulting issue.

natyork replied:
Agree with item 1 and have updated accordingly (another option could be if `errors.any?`, but i like that this is more specific to what this custom validation requires)

WRT item 2, as discussed, there is already a duration validation in the presentation model but it seems that rails will run all validations even if one fails, so if you have a custom validation that is dependent on an attribute being valid, you will have to test explicity for this in your custom validation. Ref http://stackoverflow.com/questions/5966055/controlling-the-order-of-rails-validations

jacquesporveau commented on app/views/presentations/show.html.erb:
```diff
@@ -1,3 +1,5 @@
+<%= flash[:notice] %>
```
Wonder if this should be included in the application layout?

natyork replied:
Great idea!

jacquesporveau commented on spec/controllers/presentations_controller_spec.rb:
```diff
@@ -6,10 +6,19 @@
+      it { is_expected.to be_successful }
+    end
+
+    describe 'GET #edit' do
```
You're already testing the #edit action above. Dupe test.

natyork replied:
Fixing up spec after rebase mixup, and will be reorganizing per @brendandeere's comment above.

natyork commented:
Please disregard the presentation controller spec for the time being. I messed it up with rebasing so I'm currently fixing.

natyork commented:
Updated PR per @brendandeere and @jacquesporveau comments. When refactoring the spec for the presentation controller, I rearranged a lot of code, so I did not break the individual changes for each method into the separate commits, as I normally do. This seemed to make the most sense to me in terms of ease of review, but please let me know if you'd prefer to see this broken up into separate commits!

jacquesporveau commented on app/views/presentations/show.html.erb:
```diff
@@ -10,3 +10,9 @@
+<%= link_to 'Edit', edit_presentation_path(@presentation) %>
+<%= link_to 'Delete', presentation_path(@presentation),
+  method: :delete,
+  data: { confirm: 'Are you sure?' } %>
```
I'd kind of like to see these parameters lined up a little more but if rubocop is alright with it then I am too.

natyork replied:
As discussed, rubocop only checks the .rb files so this file wouldn't have been been checked. However you bring up an excellent point - we seem to be using different indenting styles in our views to break up long lines. I just found this thoughtbot styleguide that has an opinion on this:
https://github.com/thoughtbot/guides/blob/master/style/erb/sample.html.erb

jacquesporveau commented on spec/controllers/presentations_controller_spec.rb:
```diff
@@ -49,45 +143,72 @@
 
-        it 'throws a not authorized error' do
-          expect { subject }.to raise_error(Pundit::NotAuthorizedError)
+        it 'shouldn\'t update presentation' do
```
Personally I like to avoid escaping characters in strings but this is fine. 

Some ways you can avoid escaping characters that I've found is just excluding punctuation or using double quotations as the opening and closing string limiters.

natyork replied:
Fair enough - not really sure where my head was at there ;) Now it's super bugging me, so I'm going to change it!

jacquesporveau commented on spec/controllers/presentations_controller_spec.rb:
```diff
@@ -49,45 +143,72 @@
-      subject { get :index }
+  describe 'DELETE #destroy' do
+    subject { delete :destroy, params: { id: presentation.id } }
+    let!(:presentation) { FactoryGirl.create(:presentation) }
```
Do we need a ! here? I feel like when you make the request the presentation would be created before it was sent so it would be available in the controller.

natyork replied:
As discussed, the way this test is implemented it doesn't work without the bang because when presentation gets called in the subject block it immediately destroys the presentation, and so the test to check if the presentation count has changed will fail.
